File: traffic_signaling/tests/test_time.cpp
#include <chrono>
#include <iostream>
#include <string>
#include <vector>

#include "simulation/simulation.hpp"

#define TIME(statement, name)                          \
    start = std::chrono::high_resolution_clock::now(); \
    statement;                                         \
    end = std::chrono::high_resolution_clock::now();   \
    std::cout << name << ": " << std::chrono::duration<double>{end - start}.count() << "s\n";

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    auto &&input_file = args[0];
    auto &&plan_file = args[1];

    std::cout << "------------------------------- DATA "
              // Ad hoc way to get the data name from the input file name.
              << input_file.substr(input_file.find(".txt") - 1, 1)
              << " -------------------------------\n";

    std::chrono::high_resolution_clock::time_point start, end;

    TIME(city_plan::CityPlan city_plan{input_file}, "CityPlan constructor")
    TIME(simulation::Simulation simulation{city_plan}, "Simulation constructor")
    TIME(simulation.default_schedules(), "default_schedules")
    TIME(simulation.write_schedules(plan_file), "write_schedules")
    TIME(simulation.read_schedules(plan_file), "read_schedules")
    TIME(auto score = simulation.score(), "score")

    std::cout << score << "\n";
}

File: traffic_signaling/tests/test_time.py
import os
import shutil
import time
import unittest

from parameterized import parameterized

from _resolve_imports import *

class TestTime(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        test_name = __file__.split(os.sep)[-1].removesuffix('.py')
        cls.output_dir = f'{os.path.dirname(__file__)}{os.sep}out_{test_name}'
        os.makedirs(cls.output_dir, exist_ok=True)

    @classmethod
    def tearDownClass(cls):
        shutil.rmtree(cls.output_dir)

    @parameterized.expand([
        ('a'),
        ('b'),
        ('c'),
        ('d'),
        ('e'),
        ('f')
    ])
    def test_time(self, data):
        input = get_data_filename(data)
        output = f'{self.output_dir}/{data}.txt'

        print(f'\n{"-" * 31} DATA {data} {"-" * 31}')
        start = time.time()
        city_plan = CityPlan(input)
        print(f'CityPlan constructor: {time.time() - start:.4f}s')

        start = time.time()
        simulation = Simulation(city_plan)
        print(f'Simulation constructor: {time.time() - start:.4f}s')

        start = time.time()
        simulation.default_schedules()
        print(f'default_schedules: {time.time() - start:.4f}s')

        start = time.time()
        simulation.write_schedules(output)
        print(f'write_schedules: {time.time() - start:.4f}s')

        start = time.time()
        simulation.read_schedules(output)
        print(f'read_schedules: {time.time() - start:.4f}s')

        start = time.time()
        simulation.score()
        print(f'score: {time.time() - start:.4f}s')

if __name__ == '__main__':
    unittest.main()
